### PR TITLE
Tile metadata

### PR DESCRIPTION
**Pull request by ptrgags**
This PR is another 3D Tiles Next PR that targets `3d-tiles-next`

This implements the [Tile Metadata section](https://github.com/CesiumGS/3d-tiles/tree/3d-tiles-next/extensions/3DTILES_metadata/1.0.0#tile-metadata) of the `3DTILES_metadata` 1.0.0 spec. It also integrates details #9445, so tile metadata can always be accessed via `tile.metadata.getProperty()` and similar methods.

@lilleyse could you review?

To Do:
- [x] Add some specs for the new `Check.typeof.object` in ImplicitSubtree I just added.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ptrgags!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by ptrgags** [comment]
@lilleyse updated!

File: Source/Scene/ImplicitTileMetadata.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+/**
+ * Metadata about a 3D tile, from a <code>3DTILES_metadata</code> extension
+ * within a subtree from the <code>3DTILES_implicit_tiling</code> extension
+ * <p>
+ * </p>
+ *
+ * @param {ImplicitSubtree} options.implicitSubtree The implicit subtree the tile belongs to. It is assumed that the subtree's readyPromise has already resolved.
+ * @param {ImplicitTileCoordinates} options.implicitCoordinates Implicit tiling coordinates for the tile.
+ * @param {MetadataClass} [options.class] The class that group metadata conforms to.
```
Comment (lilleyse) [inline Source/Scene/ImplicitTileMetadata.js:12]:
```suggestion
 * @param {MetadataClass} [options.class] The class that tile metadata conforms to.
```

?

File: Source/Scene/ImplicitTileMetadata.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
+import Check from "../Core/Check.js";
+import defaultValue from "../Core/defaultValue.js";
+
+/**
+ * Metadata about a 3D tile, from a <code>3DTILES_metadata</code> extension
+ * within a subtree from the <code>3DTILES_implicit_tiling</code> extension
+ * <p>
+ * </p>
```
Comment (lilleyse) [inline Source/Scene/ImplicitTileMetadata.js:8]: Remove these

File: Source/Scene/ImplicitTileMetadata.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  class: {
+    get: function () {
+      return this._class;
+    },
+  },
+
+  /**
+   * Extras in the JSON object.
+   *
+   * @memberof MetadataGroup.prototype
```
Comment (lilleyse) [inline Source/Scene/ImplicitTileMetadata.js:58]:
```suggestion
   * @memberof ImplicitTileMetadata.prototype
```

File: Source/Scene/ImplicitTileMetadata.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  extras: {
+    get: function () {
+      return this._extras;
+    },
+  },
+
+  /**
+   * Extensions in the JSON object.
+   *
+   * @memberof MetadataGroup.prototype
```
Comment (lilleyse) [inline Source/Scene/ImplicitTileMetadata.js:71]:
```suggestion
   * @memberof ImplicitTileMetadata.prototype
```

File: Source/Scene/ImplicitTileMetadata.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  return this._metadataTable.getProperty(this._entityId, propertyId);
+};
+
+/**
+ * Sets the value of the property with the given ID.
+ * <p>
+ * If the property is normalized a normalized value must be provided to this function.
+ * </p>
+ * <p>
+ * If a property with the given ID doesn't exist, it is created.
```
Comment (lilleyse) [inline Source/Scene/ImplicitTileMetadata.js:121]: > If a property with the given ID doesn't exist, it is created

This applies to `TilesetMetadata` and `GroupMetadata` but doesn't apply to anything that uses metadata tables. Maybe we should remove this capability from everywhere until we figure out a consistent solution.

Also we're a bit inconsistent about the level of detail in the doc: `MetadataTable.prototype.setProperty` has more detail than `TilesetMetadata`/`GroupMetadata`.

In general we should be more consistent with these different types of metadata. Most of that is on me from the initial set of PRs.

File: Source/Scene/TileMetadata.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
+import Check from "../Core/Check.js";
+import defaultValue from "../Core/defaultValue.js";
+import MetadataEntity from "./MetadataEntity.js";
+
+/**
+ * Metadata about a 3D tile. This can either be defined directly with
+ * a <code>3DTILES_metadata</code> on the tile, or indirectly by using
+ * <code>3DTILES_metadata</code> inside an implicit subtree file.
+ * <p>
+ * options.tile and options.implicitSubtree are mutually exclusive.
```
Comment (lilleyse) [inline Source/Scene/TileMetadata.js:10]: Outdated doc?

File: Source/Scene/TileMetadata.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+ * </p>
+ *
+ * @param {Object} options Object with the following properties:
+ * @param {Object} [options.tile] The extension JSON attached to the tile.
```
Comment (lilleyse) [inline Source/Scene/TileMetadata.js:14]:
```suggestion
 * @param {Object} options.tile The extension JSON attached to the tile.
```

File: Source/Scene/TileMetadata.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+   */
+  class: {
+    get: function () {
+      return this._class;
+    },
+  },
+
+  /**
+   * A dictionary containing properties. This is only defined when
+   * explicit tile metadata is used.
```
Comment (lilleyse) [inline Source/Scene/TileMetadata.js:51]: Outdated doc?

File: Source/Scene/TileMetadata.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+  properties: {
+    get: function () {
+      return this._properties;
+    },
+  },
+
+  /**
+   * Extras in the JSON object.
+   *
+   * @memberof MetadataGroup.prototype
```
Comment (lilleyse) [inline Source/Scene/TileMetadata.js:66]:
```suggestion
   * @memberof TileMetadata.prototype
```

File: Source/Scene/TileMetadata.js [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+  extras: {
+    get: function () {
+      return this._extras;
+    },
+  },
+
+  /**
+   * Extensions in the JSON object.
+   *
+   * @memberof MetadataGroup.prototype
```
Comment (lilleyse) [inline Source/Scene/TileMetadata.js:79]:
```suggestion
   * @memberof TileMetadata.prototype
```

File: Source/Scene/ImplicitTileMetadata.js [hunk 11]
```diff
@@ -0,0 +1,152 @@
+import Check from "../Core/Check.js";
+import defaultValue from "../Core/defaultValue.js";
+
+/**
+ * Metadata about a 3D tile, from a <code>3DTILES_metadata</code> extension
+ * within a subtree from the <code>3DTILES_implicit_tiling</code> extension
```
Comment (lilleyse) [inline Source/Scene/ImplicitTileMetadata.js:6]: Might want to include a sentence that says this mirrors `TileMetadata`.

File: Source/Scene/Cesium3DTile.js [hunk 12]
```diff
@@ -32,6 +32,7 @@ import Cesium3DTileRefine from "./Cesium3DTileRefine.js";
 import Empty3DTileContent from "./Empty3DTileContent.js";
 import findGroupMetadata from "./findGroupMetadata.js";
 import has3DTilesExtension from "./has3DTilesExtension.js";
+import TileMetadata from "./TileMetadata.js";
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTile.js:35]: Move into alphabetic order. I'm guessing this used to be called `MetadataTile`.

File: Source/Scene/Cesium3DTile.js [hunk 13]
```diff
@@ -290,6 +291,28 @@ function Cesium3DTile(tileset, baseResource, header, parent) {
    */
   this.hasMultipleContents = hasMultipleContents;
 
+  var metadata;
+  if (has3DTilesExtension(header, "3DTILES_metadata")) {
+    // This assumes that tileset.metadata has been created before any
+    // tiles are constucted.
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTile.js:297]:
```suggestion
    // tiles are constructed.
```

File: Source/Scene/Cesium3DTile.js (hunk 12 above)
Comment (ptrgags) [reply Source/Scene/Cesium3DTile.js:35]: yup exactly, after the rename in the other PR I did a search and replace.

**Comment by ptrgags** [comment]
@lilleyse updated!

**Comment by lilleyse** [comment]
I pushed a commit to change some of the `@exception` wording so it more closely matches the printed developer error (though it's not exactly the same still). Also there was one test missing `toThrowDeveloperError` that I pushed a fix for. https://github.com/CesiumGS/cesium/pull/9464/commits/86b21cd196d676e7aa05c4c3c0fb42beef780922

**Comment by lilleyse** [comment]
Thanks @ptrgags!

**Pull request merged**